extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java

```diff
@@ -166,14 +169,14 @@ private void sendToKafka(final String topic, MemoryBoundLinkedBlockingQueue<Stri
   public void emit(final Event event)
   {
     if (event != null) {
-      ImmutableMap.Builder<String, Object> resultBuilder = ImmutableMap.<String, Object>builder().putAll(event.toMap());
-      if (config.getClusterName() != null) {
-        resultBuilder.put("clusterName", config.getClusterName());
-      }
-      Map<String, Object> result = resultBuilder.build();
-
       try {
-        String resultJson = jsonMapper.writeValueAsString(result);
+        EventMap map = event.toMap();
+        if (config.getClusterName() != null) {
+          map.put("clusterName", config.getClusterName());
```

Comment:
the map should be copied to a new structure before modification - it is better to not expect the object to be mutable

Reply:
I can replace EventMap to extend the ImmutableMap instead of HashMap, what do you think? Then it will be always immutable on the output.

Reply:
If that works, I think that'll be ok. But I'm not sure if it will expose all methods which `HashMap` does. Maybe, can also return an `ImmutableMap` object from the builder's build. Immutability would also require making the default constructor private so that we don't create a mutable map from there.

Reply:
If we return ImmutableMap then we get back to the serialization problem with Map and polymorphic types in Jackson. `build` needs to return EventMap in this case.

Reply:
Ah, yes you're right. In that case, maybe you need to create an `EventMap` copy here and put additional attributes in it.

Reply:
Extending ImmutableMap is tricky (and IMO not worth it). I just added this:
```java
  /**
   * Convert this EventMap to a builder. Performs copy of the whole EventMap.
   */
  public Builder asBuilder() {return new Builder().putAll(this);}
```
So you can get a builder out of the existing EventMap and modify the copy.

Reply:
And then we can use it like this:
```java
        EventMap map = event.toMap()
                            .asBuilder()
                            .putNonNull("clusterName", config.getClusterName())
                            .build();
```
or this (we avoid copying if we don't need to modify the map): 
```java
        EventMap map = event.toMap();
        if (config.getClusterName() != null) {
          map = map.asBuilder()
                   .put("clusterName", config.getClusterName())
                   .build();
        }
```

Reply:
good idea! :+1: